aten/src/ATen/native/sparse/SparseFactories.cpp

```diff
@@ -0,0 +1,176 @@
+  int64_t actual_nnz = 0;
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+      at::ScalarType::BFloat16,
+      at::ScalarType::Half,
+      at::ScalarType::Bool,
+      at::ScalarType::ComplexHalf,
+      values.scalar_type(),
+      "spdiags",
+      [&] {
+        auto values_accessor = values.accessor<scalar_t, 1>();
+        auto const diagonals_accessor = diagonals.accessor<scalar_t, 2>();
+
+        for (const auto d_i : c10::irange(n_diag)) {
+          const int64_t& off_i = offsets_accessor[d_i];
+          const int64_t row_out_begin = off_i < 0 ? std::abs(off_i) : 0;
+          const int64_t col_out_begin = off_i > 0 ? off_i : 0;
+
+          // Number of column, row positions we can assign into and the number
+          // of diag positions we can read from
+          const int64_t row_slots = n_rows_out - row_out_begin;
+          const int64_t col_slots = n_cols_out - col_out_begin;
+          const int64_t diag_slots = n_cols_in - col_out_begin;
+
+          // Max number of reads lets us limit the loop by the first thing we
+          // will exhaust, note if any of these comes up negative irange will be
+          // empty so we don't have to handle that case explicitly
+          const int64_t max_read =
+              std::min(diag_slots, std::min(row_slots, col_slots));
+
+
+          for (const auto read_count : c10::irange(max_read)) {
+            const auto col = col_out_begin + read_count;
+            const auto row = row_out_begin + read_count;
+
+            values_accessor[actual_nnz] = diagonals_accessor[d_i][col];
+            indices_accessor[0][actual_nnz] = row;
+            indices_accessor[1][actual_nnz] = col;
+            ++actual_nnz;
+          }
+        }
+      });
+  indices = indices.narrow(1, 0, actual_nnz);
+  values = values.narrow(0, 0, actual_nnz);
+  SparseTensor sparse =
+      at::sparse_coo_tensor(indices, values, shape, sparse_options);
+  return sparse;
+}
+
+// Check offsets for duplicates, and out-of-bounds diagonals
+void validate_spdiags_offsets_cpu(
+    const Tensor& offsets,
+    const int64_t n_row,
+    const int64_t n_col) {
+  std::set<int64_t> seen_offsets;
+  auto offsets_accessor = offsets.accessor<int64_t, 1>();
+  for (auto i : c10::irange(offsets.size(0))) {
+    auto off = offsets_accessor[i];
+    TORCH_CHECK(
+        seen_offsets.insert(off).second,
+        "Offset array contains duplicate values");
+    TORCH_CHECK(
+        ((-1 * n_row) < off) && (off < n_col),
+        "Diagonal ",
+        off,
+        " does not exist in output shape (",
+        n_row,
+        ",",
+        n_col,
+        ")");
+  }
+}
+
+} // namespace
+
+SparseTensor spdiags_sparse_cpu(
+    const Tensor& diagonals,
+    const Tensor& offsets,
+    IntArrayRef shape,
+    c10::optional<Layout> layout) {
+  TORCH_CHECK(diagonals.dim() == 2, "Diagonals must be 2d");
+  TORCH_CHECK(shape.size() == 2, "Output shape must be 2d");
+  TORCH_CHECK(offsets.dim() == 1, "Offsets must be 1d");
```

Comment:
Not SciPy conistent. The doc allows it to be a scalar, so `dim() == 0` should be supported for consistency.

Reply:
I don't know if that is such a bad thing, we have already departed from SciPy consistency. 

- SciPy does not require that the offsets be an array either a scalar value is accepted there. 
- SciPy requires the output shape as two separate args (we are requiring it be passed as a tuple)

Reply:
`dim = 0` is an easy case to support. If your code relies on inputs with `dim() == 1`, then `unsqueeze` will solve this issue. Anyway, embedding a single diagonal as a sparse tensor is a neat functionality which is also very easy to support.